nntrainer/layers/conv2d_layer.cpp

```diff
@@ -183,80 +164,83 @@ static void im2col(const Tensor &in, const TensorDim &kdim,
-          }
-
-          unsigned int im_w = base_im_w;
-          for (int ws = -pw; ws <= w_stride_end; ws += mstride[1]) {
-            unsigned int im_h = base_im_h;
-            int patch_width_end = eff_k_width + ws;
-
-            for (int w = ws; w < patch_width_end; w += dilation[1]) {
-              if (w < 0 || in_width <= w) {
-                im_h++;
-                continue;
-              }
-              out_data[im_w * owidth + im_h] = in.getValue(0, c, h, w);
-              im_h++;
-            }
-            im_w++;
-          }
+  // if (channel_mode) {
+  int h_stride_end = height - eff_k_height - ph;
+  int w_stride_end = width - eff_k_width - pw;
+
+  /// get a patch, size of kernel
+  /// hs is height_strided, ws is width_strided
+  unsigned int owidth = out.width();
+  unsigned int base_im_w = 0;
+  for (int hs = -ph; hs <= h_stride_end; hs += mstride[0]) {
+    unsigned int base_im_h = 0;
+    int patch_height_end = eff_k_height + hs;
+    /// map the patch to a single line looping through channel
+    for (unsigned int c = 0; c < channel; ++c) {
+      for (int h = hs; h < patch_height_end; h += dilation[0]) {
+        if (h < 0 || in_height <= h) {
           base_im_h += k_width;
+          continue;
         }
-      }
-      base_im_w += out_width;
-    }
-  } else {
-    unsigned int im_w = 0;
-
-    if (eff_k_height > height || eff_k_width > width)
-      throw std::runtime_error("Kernel shape bigger than input shape");
-
-    for (unsigned int c = 0; c < channel; ++c) {
-      for (unsigned int hs = 0; hs <= height - eff_k_height; hs += mstride[0]) {
-        for (unsigned int ws = 0; ws <= width - eff_k_width; ws += mstride[1]) {
-          unsigned int im_h = 0;
-          unsigned int patch_height_end = eff_k_height + hs;
-          unsigned int patch_width_end = eff_k_width + ws;
-
-          for (unsigned int h = hs; h < patch_height_end; h += dilation[0]) {
-            if (h < ph || in_height + ph <= h) {
-              im_h += k_width;
-              continue;
-            }
 
-            for (unsigned int w = ws; w < patch_width_end; w += dilation[1]) {
-              if (w < pw || in_width + pw <= w) {
-                im_h++;
-                continue;
-              }
+        unsigned int im_w = base_im_w;
+        for (int ws = -pw; ws <= w_stride_end; ws += mstride[1]) {
+          unsigned int im_h = base_im_h;
+          int patch_width_end = eff_k_width + ws;
 
-              float val = in.getValue(0, c, h - ph, w - pw);
-              out.setValue(0, 0, im_h, im_w, val);
+          for (int w = ws; w < patch_width_end; w += dilation[1]) {
+            if (w < 0 || in_width <= w) {
               im_h++;
+              continue;
             }
+            out_data[im_w * owidth + im_h] = in.getValue(0, c, h, w);
+            im_h++;
           }
           im_w++;
         }
+        base_im_h += k_width;
       }
     }
+    base_im_w += out_width;
   }
+  // } else {
```

Comment:
Is this intentional?

Reply:
Yes, because im2col is in the critical path, But I wanted to leave the code so commented away.

Reply:
So, the commented code is also left for transposed convolution?

Reply:
Yup.

Reply:
Then please remove the `if else` statements and make a comment at top of this code for the reason this is left.

Reply:
There is two if else statement so I think it's better to saperate the logic out of the function, let me handle this later.